### ISSUE TITLE
Problems with updateData

### ISSUE THREAD

**Issue by tomheaps** [olifolkerd/tabulator#625]
I am getting the following error message in the console when I use the updateData function to update my table using an ajax request via PHP to a MySQL database:

```
tabulator.min.js:3 Uncaught TypeError: t.forEach is not a function
    at t.(anonymous function).(anonymous function).updateData 
```

I am returning the data correctly to tabulator as JSON encoded array of objects - I know that the data is being returned properly because when I use setData instead of updateData (with exactly the same ajax request and PHP file) everything works perfectly and the data from my database is returning correctly - I don't want to keep calling setData however as this is extremely memory intensive and eventually causes the browser to crash.

Any ideas why updateData is not working whereas setData is?

**Comment by olifolkerd** [comment]
Hey,

I will have a look into the ***updateData*** function now.

In the mean time, the ***setData*** function shouldn't be memory intensive, have you set the ***height*** property of the table in your constructor to enable the Virtual DOM. With the virtual DOM enabled Tabulator can handle any size of data array (even into the 100,000's of rows) without any performance hit.

I will get back to you with an update on the updateData function shortly

**Comment by olifolkerd** [comment]
I have had a look at the ***updateData***function and it appears to be working correctly.

Are you sure you are passing in an array to the ***updateData*** function rather than an array as a JSON encoded string?

The ***setData*** function will accept any of the following:
- Array
- string of a JSON encoded array
- URL

The ***updateData*** function will only accept an array. The error you are getting suggest that the data being passed into the function is not an array.

Let me know if that helps

Cheers

Oli

**Comment by tomheaps** [comment]
Hi,

Please forgive my incompetence. I have little experience using ajax/PHP to return data.
When I log the data returned by my ajax request in the console I get:

```
Array(36)
{id: "C1", ward: "AMU", name: "", time: "", elapsed: "", …}
{id: "A4", ward: "AMU", name: "DSGDGSDGSD, Gdsgdsg", time: "17 10 07 21 47", elapsed: 1477, …}
{id: "SR1", ward: "AMU", name: "", time: "", elapsed: "", …}
{id: "A3", ward: "AMU", name: "", time: "", elapsed: "", …}
{id: "C4", ward: "AMU", name: "", time: "", elapsed: "", …}
{id: "A6", ward: "AMU", name: "", time: "", elapsed: "", …}
{id: "B10", ward: "AMU", name: "", time: "", elapsed: "", …}
{id: "B2", ward: "AMU", name: "", time: "", elapsed: "", …}
{id: "B3", ward: "AMU", name: "", time: "", elapsed: "", …}
etc. etc.
```

The PHP code I am using to return data is:

```
<?php
require '../../../db_config.php';

    $sql = "select * from AMU";
    $result = mysqli_query($connection, $sql) or die("Error in Selecting " . mysqli_error($connection));

    $emparray = array();
    while($row =mysqli_fetch_assoc($result))
    {
        $emparray[] = $row;
    }
    echo json_encode($emparray);

    mysqli_close($connection);
?>
```

From what you are saying, my PHP is probably returning a string of a JSON encoded array?

If I am correct, how should I write my php to return an array that is compatible with the updateData function?

Thanks so much for your help again.

**Comment by olifolkerd** [comment]
Your PHP should always return it as a JSON encoded array, that is the best way to send it over in a request, it is how things are handled on the client side that matters. 

What shows up when you do the following console log, (you want to replace the ***your_var*** variable with the variable you are passing to the ***updateData*** function):
```js
console.log(typeof your_var);
```

**Comment by tomheaps** [comment]
Hi, the console log states 'object'

**Comment by olifolkerd** [comment]
what happens when you try:
```js
console.log(your_var[0]);
```

**Comment by olifolkerd** [comment]
which version to tabulator are you using, and on what browser?

**Comment by tomheaps** [comment]
latest version of tabulator, 3.3 on chrome v61

**Comment by olifolkerd** [comment]
could you post a copy of your table constructor

**Comment by olifolkerd** [comment]
it looks like you are passing in an array, i will need to see your table constructor to test things more thoroughly

**Comment by olifolkerd** [comment]
also if you could post the updateData function you are using as well that would be a big help

**Comment by tomheaps** [comment]
when I try `console.log(data[0]);` I get `{id: "C1", ward: "AMU", name: "", time: "", elapsed: "", …}`

**Comment by olifolkerd** [comment]
that looks right
=

**Comment by tomheaps** [comment]
I know but when I do the following with my ajax request:

```
success: function(data) {
      if (data.isArray) {
        console.log("array")
      } else {
        console.log("not array")
      }
```

The console returns "not array"????

**Comment by olifolkerd** [comment]
what about:
```js
console.log(Array.isArray(data));
```

**Comment by olifolkerd** [comment]
could you post a copy of your table constructor and your updateData function call

**Comment by tomheaps** [comment]
Here is all my table construct and as I said earlier, the setData function at the bottom calls on the same PHP file with the same config options and works fine however the updateData function contained within my setInterval(updateTable, 1000) is broken :(

```

  //Editors and Formatters for table-------------------------------------------------------------------

  var editCheck = function(cell) {
    var data = cell.getRow().getData();
    return data.name !== ""
  }

  var specialtyEditor = function(cell, onRendered, success, cancel) {
    var editor = $("<select class='form-control'><option value=''></option><option value='Cardiology'>Cardiology</option><option value='Frailty'>Frailty</option><option value='Orthopaedics'>Orthopaedics</option><option value='Respiratory'>Respiratory</option><option value='General Surgery'>General Surgery</option></select>");
    editor.val(cell.getValue());
    onRendered(function() {
      editor.focus();
    });
    editor.on("change blur", function(e) {
      success(editor.val());
    });
    return editor;
  };

  var outcomeEditor = function(cell, onRendered, success, cancel) {
    var editor = $("<select class='form-control'><option value=''></option><option value='Potential Discharge'>? Discharge</option><option value='Definite Discharge'>Discharge</option><option value='AMU'>Stay on AMU</option><option value='Any Medical'>Any Medical</option><option value='Any CoE'>Any CoE</option><option value='Avon'>Avon</option><option value='Castle'>Castle</option><option value='CCU'>CCU</option><option value='Fairfax'>Fairfax</option><option value='Farries'>Farries</option><option value='ITU'>ITU</option><option value='Malins'>Malins</option><option value='Mary'>Mary</option><option value='Nicholas'>Nicholas</option><option value='Victoria'>Victoria</option></select>");
    editor.val(cell.getValue());
    onRendered(function() {
      editor.focus();
    });
    editor.on("change blur", function(e) {
      success(editor.val());
    });
    return editor;
  }

  var allocatedEditor = function(cell, onRendered, success, cancel) {
    var editor = $("<select class='form-control'><option value=''></option><option value='Avon'>Avon</option><option value='Castle'>Castle</option><option value='CCU'>CCU</option><option value='Fairfax'>Fairfax</option><option value='Farries'>Farries</option><option value='ITU'>ITU</option><option value='Malins'>Malins</option><option value='Mary'>Mary</option><option value='Nicholas'>Nicholas</option><option value='Victoria'>Victoria</option></select>");
    editor.val(cell.getValue());
    onRendered(function() {
      editor.focus();
    });
    editor.on("change blur", function(e) {
      success(editor.val());
    });
    return editor;
  }

  var doneNotdone = function(cell, formatterParams) {
    var cellValue = cell.getValue();
    if (cellValue == "true") {
      return "<i class='green fal fa-check'></i>"
    } else if (cellValue == "false") {
      return "<i class='red fal fa-times'></i>"
    }
  };

  var checkBox = function(cell, formatterParams) {
    var cellValue = cell.getValue();
    if (cellValue == "done") {
      return "<i class='green fa fa-check-circle'></i>"
    } else if (cellValue == "yes") {
      return "<i class='red fa fa-circle'></i>"
    } else if (cellValue === "") {
      return ""
    }
  };

  var checkBox2 = function(cell, formatterParams) {
    var cellValue = cell.getValue();
    if (cellValue == "done") {
      return "<i class='green fa fa-check-circle'></i>"
    } else if (cellValue == "yes") {
      return "<i class='orange fa fa-circle'></i>"
    } else if (cellValue == "urgent") {
      return "<i class='red fa fa-circle'></i>"
    }
  };

  var thinkGlucose = function(cell, formatterParams) {
    var cellValue = cell.getValue();
    if (cellValue == "true") {
      return '<span class="fa-layers" style="font-size:1rem"><i class="red fa fa-circle"></i><span class="fa-layers-text fa-inverse" data-fa-transform="shrink-5">D</span></span>'
    }
    if (cellValue == "false") {
      return ""
    }
  };

  var outcomeFormatter = function(cell, formatterParams) {
    var cellValue = cell.getValue();
    if (cellValue == "AMU") {
      return "<span style='color:#fa5252;font-family:Suisse-Black;'>" + cellValue + "</span>"
    } else if (cellValue == "Definite Discharge") {
      return "<i class='green fa fa-home'></i>"
    } else if (cellValue == "Potential Discharge") {
      return "<i class='orange fa fa-home'></i>"
    } else {
      return cellValue
    }
  };

  var timeElapsed = function(value, data, type, mutatorParams) {
    var timeCol = data.time;
    var startTime = moment(timeCol, "YY MM DD HH mm");
    var currentTime = moment();
    var elapsedTime = currentTime.diff(startTime, 'minutes');
    if (timeCol === "") {
      return ""
    } else {
      return elapsedTime
    }
  }

  var formatLOS = function(cell, formatterParams) {
    var cellValue = cell.getValue();
    var elapsedCol = cell.getData().elapsed
    if (elapsedCol === "") {
      return ""
    } else if ((elapsedCol > 1339) && (elapsedCol < 2880)) {
      cell.getElement().css('color', '#fcc419');
      return cellValue;
    } else if (elapsedCol > 2879) {
      cell.getElement().css('color', '#fa5252');
      return cellValue;
    } else {
      return cellValue
    }
  }

  var mutateLOS = function(value, data, type, mutatorParams) {
    var elapsedTime = data.elapsed;
    var los = moment.duration(elapsedTime, "minutes").format("d[d] h[h] m[m]");
    if (elapsedTime === "") {
      return ""
    } else {
      return los
    }
  }

  var deleteIcon = function(cell, formatterParams) {
    return "<i class='red fal fa-minus-circle'></i>";
  };
  var admitIcon = function(cell, formatterParams) {
    return "<i class='blue fal fa-plus-circle'></i>";
  };

  var moveIcon = function(cell, formatterParams) {
    return "<i class='blue fal fa-arrows-v'></i>";
  };

  function saveData() {
    var data = $("#amu-whiteboard").tabulator("getData");
    localStorage.setItem("tabulator-data", JSON.stringify(data));
  }

  function statusUpdate() {
    var empty = $('[tabulator-field="name"]:empty').length;
    var sr1 = $('.tabulator-row:nth-of-type(29) [tabulator-field="name"]:empty').length;
    var sr2 = $('.tabulator-row:nth-of-type(30) [tabulator-field="name"]:empty').length;
    var b1 = $('.tabulator-row:nth-of-type(7) [tabulator-field="name"]:empty').length;
    var b2 = $('.tabulator-row:nth-of-type(8) [tabulator-field="name"]:empty').length;
    var c10 = $('.tabulator-row:nth-of-type(27) [tabulator-field="name"]:empty').length;
    var c11 = $('.tabulator-row:nth-of-type(28) [tabulator-field="name"]:empty').length;
    var emptySR = sr1 + sr2 + b1 + b2 + c10 + c11;
    var emptyBeds = empty - emptySR;
    $('#emptyBeds').text(emptyBeds);
    $('#emptySR').text(emptySR);
    var avg = 0;
    var amount = 0;
    var elapsedData = $('[tabulator-field="elapsed"]:not(:empty)');
    elapsedData.each(function() {
      avg += +($(this).text());
      amount++;
    });
    avg /= amount;
    var avLOS = moment.duration(avg, "minutes").format("d[d] h[h] m[m]");
    $("#avLOS").text(avLOS);
    var defDisch = $('.green.fa-home').length;
    var potDisch = $('.orange.fa-home').length;
    var discharges = defDisch + potDisch;
    $('#discharges').text(discharges);
    $('#potDisch').text(potDisch);
  }

  function sortByBed() {
    $("#amu-whiteboard").tabulator("setSort", "id", "asc");
  }

  function checkDuplicates() {
    var index = {},
      getText = function() {
        return $.trim($(this).text().split(', ')[0]);
      };
    $("#amu-whiteboard").find(".tabulator-row").each(function() {
      var rowKey = $(this).find('.tabulator-cell[tabulator-field="name"]').map(getText).toArray().join("|");
      if (index.hasOwnProperty(rowKey)) {
        index[rowKey].push(this);
      } else {
        index[rowKey] = [this];
      }
    });
    $.each(index, function(rowKey, rows) {
      $(rows).toggleClass("duplicate", rows.length > 1);
    });
  }
  var ajaxConfig = {
    success: function(data) {
      if (data) {
        console.log(data)
      }
    },
    complete: function() {
      sortByBed();
      statusUpdate();
      checkDuplicates();
    }
  };

  function updateTable() {
    $("#amu-whiteboard").tabulator("updateData", "php/amu-fetch.php", {}, ajaxConfig);
  }

  //Build Table-------------------------------------------------------------------

  $("#amu-whiteboard").tabulator({
    layout: "fitColumns",
    resizableColumns: false,
    groupBy: "ward",
    groupToggleElement: "header",
    index: "id",
    initialSort: [{
      column: "id",
      dir: "asc"
    }],
    ajaxLoaderLoading: "Updating Data",
    ajaxLoaderError: "Error Updating Data </br> Data will be lost if you log off!",
    columns: [{
        title: "Bed",
        field: "id",
        align: "center",
        headerSort: false,
        width: 80,
        sorter: "alphanum",
      }, {
        title: "Ward",
        field: "ward",
        align: "center",
        headerSort: false,
        width: 80,
        editor: true,
        visible: false,
      },
      {
        title: "Patient Name",
        field: "name",
        width: 300,
        sorter: "string",
        cellClick: function(e, cell) {
          if (cell.getValue() === "") {
            return false
          } else {
            var row = cell.getRow()
            var bed = row.getIndex();
            var name = cell.getValue();
            var fullName = name.split(', ');
            var lastName = fullName[0];
            var firstName = fullName[fullName.length - 1];
            $('#bedNo').text(bed);
            $('#forenameEdit').val(firstName);
            $('#surnameEdit').val(lastName);
            $('#editModal').modal('show');
          }
        }
      }, {
        title: "Time",
        field: "time",
        align: "center",
        editor: true,
        visible: false,
      }, {
        title: "Elapsed",
        field: "elapsed",
        mutator: timeElapsed,
        width: 120,
        align: "center",
        sorter: "number",
        editor: true,
        visible: false,
      }, {
        title: "LOS",
        field: "los",
        align: "center",
        width: 120,
        mutator: mutateLOS,
        formatter: formatLOS,
        sorter: function(a, b, aRow, bRow, sorterParams) {
          return (aRow.getData().elapsed) - (bRow.getData().elapsed);
        }
      }, {
        title: "Specialty",
        field: "specialty",
        align: "center",
        width: 140,
        editable: editCheck,
        editor: specialtyEditor,
        formatter: function(cell, formatterParams) {
          var cellValue = cell.getValue();
          if (cellValue == "General Surgery") {
            cell.getRow().getElement().css({
              "background-color": "rgba(250,82,82,0.2)"
            });
            return cellValue;
          } else if (cellValue == "Respiratory") {
            cell.getRow().getElement().css({
              "background-color": "rgba(255, 212, 59, 0.2)"
            });
            return cellValue;
          } else if (cellValue == "Frailty") {
            cell.getRow().getElement().css({
              "background-color": "rgba(105, 219, 124, 0.2)"
            });
            return cellValue;
          } else if (cellValue == "Orthopaedics") {
            cell.getRow().getElement().css({
              "background-color": "rgba(250,82,82,0.2)"
            });
            return cellValue;
          } else if (cellValue == "Cardiology") {
            cell.getRow().getElement().css({
              "background-color": "rgba(28, 124, 214, 0.2)"
            });
            return cellValue;
          } else {
            cell.getRow().getElement().css({
              "background-color": "transparent"
            });
            return cellValue;
          }
        }
      }, {
        title: "DM",
        field: "diabetes",
        cellClick: function(e, cell) {
          var cellValue = cell.getValue();
          var row = cell.getRow();
          var name = row.getCell("name").getValue();
          if (name === "") {
            return false
          } else {
            if (cellValue == "true") {
              cell.setValue("false")
            } else {
              cell.setValue("true")
            }
          }
        },
        formatter: thinkGlucose,
        align: "center",
        width: 80,
      }, {
        title: "Meds",
        field: "medrec",
        cellClick: function(e, cell) {
          var cellValue = cell.getValue();
          if (cellValue == "urgent") {
            cell.setValue("done")
          } else if (cellValue == "yes") {
            cell.setValue("done")
          } else if (cellValue == "done") {
            cell.setValue("yes")
          }
        },
        cellDblClick: function(e, cell) {
          var cellValue = cell.getValue();
          if (cellValue == "yes") {
            cell.setValue("urgent")
          } else {
            return false
          }
        },
        formatter: checkBox2,
        align: "center",
        width: 110
      },
      {
        title: "PT",
        field: "pt",
        cellClick: function(e, cell) {
          var cellValue = cell.getValue();
          var row = cell.getRow();
          var name = row.getCell("name").getValue();
          if (name === "") {
            return false
          } else {
            if (cellValue === "") {
              cell.setValue("yes")
            } else if (cellValue == "yes") {
              cell.setValue("done")
            } else if (cellValue == "done") {
              cell.setValue("")
            }
          }
        },

        formatter: checkBox,
        align: "center",
        width: 110
      }, {
        title: "OT",
        field: "ot",
        cellClick: function(e, cell) {
          var cellValue = cell.getValue();
          var row = cell.getRow();
          var name = row.getCell("name").getValue();
          if (name === "") {
            return false
          } else {
            if (cellValue === "") {
              cell.setValue("yes")
            } else if (cellValue == "yes") {
              cell.setValue("done")
            } else if (cellValue == "done") {
              cell.setValue("")
            }
          }
        },

        formatter: checkBox,
        align: "center",
        width: 110
      }, {
        title: "Outcome",
        field: "outcome",
        align: "center",
        editor: outcomeEditor,
        editable: editCheck,
        formatter: outcomeFormatter,
        width: 160
      }, {
        title: "Allocated",
        field: "allocated",
        editor: allocatedEditor,
        align: "center",
        width: 160,
        editable: editCheck,
      }, {
        title: "Notes / Plan",
        field: "notes",
        editor: true,
        editable: editCheck,
        headerSort: false,
      }, {
        title: "Move",
        width: 70,
        formatter: moveIcon,
        align: "center",
        headerSort: false,
        cellClick: function(e, cell) {
          var row = cell.getRow()
          var bed = row.getIndex();
          var name = row.getCell("name").getValue();
          var ward = row.getCell("ward").getValue();
          var moveModal = $('#moveModal');
          if (name === "") {
            return false
          } else {
            var bedSelect = $('#bedSelect');
            $('#moveBed, #moveBed2').text(bed);
            $('#moveName, #moveName2').text(name);
            $("#bedSelect option").prop("selected", false).filter(function() {
              return $.trim($(this).text()) == bed;
            }).prop('selected', true);
            moveModal.modal('show');
            $('#moveButton').off().on('click', function(e) {
              var newBed = bedSelect.val();
              var newRow = $('#amu-whiteboard').tabulator("getRow", newBed);
              var newWard = newRow.getData().ward;
              console.log(ward + newWard)
              $.ajax({
                data: 'idstart=' + bed + '&idnew=' + newBed + '&ward=' + ward + '&newward=' + newWard,
                url: "php/amu-move.php",
                type: "POST",
                complete: function() {
                  updateTable();
                  sortByBed();
                  saveData();
                  $.ajax({
                    data: 'idstart=' + bed + '&idnew=' + newBed + '&ward=' + ward + '&newward=' + newWard,
                    url: "php/amutasks-move.php",
                    type: "POST",
                  })
                }
              });
              moveModal.modal('hide');

            })
          }
        }
      },
      {
        title: "Admit",
        width: 70,
        formatter: admitIcon,
        align: "center",
        headerSort: false,
        cellClick: function(e, cell) {
          var row = cell.getRow()
          var bed = row.getIndex();
          var name = row.getCell("name").getValue();
          var addModal = $('#addModal');
          if (name === "") {
            $('#admitBed').text(bed);
            addModal.modal('show');
          } else {
            $('#addwarnBed').text(bed);
            $('#addwarnName').text(name);
            $('#addwarnModal').modal('show');
          }
        }
      }, {
        title: "Disch",
        width: 70,
        formatter: deleteIcon,
        align: "center",
        headerSort: false,
        cellClick: function(e, cell) {
          var row = cell.getRow()
          var name = row.getCell("name").getValue();
          var deleteModal = $('#deleteModal');
          var bed = row.getIndex();

          if (name === "") {
            return false
          } else {
            $('#dischargeName').text(name);
            $('#dischargeBed').text(bed);
            deleteModal.modal('show');
            deleteModal.on('click', '.button-cancel', function() {
              deleteModal.modal('hide');
            });
            deleteModal.on('click', '.button-warning', function() {
              row.update({
                "name": "",
                "time": "",
                "los": "",
                "specialty": "",
                "diabetes": "",
                "medrec": "",
                "pt": "",
                "ot": "",
                "outcome": "",
                "allocated": "",
                "notes": ""
              });
              deleteModal.modal('hide');
            });
          }
        }
      }
    ],
    dataEdited: function(data) {
      localStorage.setItem("tabulator-data", JSON.stringify(data));
    },
    rowUpdated: function(row) {
      var data = row.getData();
      var id = row.getIndex();
      var name = data.name
      $.ajax({
        data: data,
        dataType: "json",
        type: "POST",
        url: "php/amu-add-delete.php",
        complete: function() {
          checkDuplicates();
          sortByBed();
          saveData();
          if (name === undefined || name === "") {
            $.ajax({
              data: 'id=' + id,
              type: "POST",
              url: "php/amutasks-delete.php"
            })
          } else {
            return false
          }
        }
      });



    },
    cellEdited: function(cell) {
      var row = cell.getRow();
      var value = cell.getValue();
      var field = cell.getField();
      var bed = row.getIndex();
      $.ajax({
        data: 'field=' + field + '&value=' + value + '&id=' + bed,
        url: "php/amu-update.php",
        type: "POST",
        complete: function() {
          if (field === "specialty") {
            $.ajax({
              data: 'value=' + value + '&id=' + bed,
              type: "POST",
              url: "php/amutasks-specialty.php"
            })
          } else {
            return false
          }
        }
      })
    },
    ajaxError: function(xhr, textStatus, errorThrown) {
      var data = localStorage.getItem("tabulator-data");
      if (data) {
        $("#amu-whiteboard").tabulator("setData", JSON.parse(data));
      }
    },
  });
  $('#amu-whiteboard').tabulator("setData", "php/amu-fetch.php", {}, ajaxConfig);




  $(window).resize(function() {
    $("#amu-whiteboard").tabulator("redraw", true);
  });

  setInterval(updateTable, 10000);

```

**Comment by olifolkerd** [comment]
I see your problem, 

You are passing a URL to the ***updateData*** function, as i mentioned above, while the ***setData*** function can take a URL and load the data from it, the ***updateData*** data function can only take an array, it cannot lookup the data from a URL.

You will need to make the ajax request yourself and pass the returned data to the function:
```js
$.ajax({
	url: "php/amu-fetch.php",
	type: "get",
	async: true,
	dataType:'json',
	success: function (data) {
		$("#amu-whiteboard").tabulator("updateData", data);
	},
});
```

Cheers

Oli

**Comment by tomheaps** [comment]
Thanks so much, update working now however causing serious performance issues with memory bloat of page to up to 1GB within 10 minutes!!

Will have to call it a night and have a fresh look tomorrow.

**Comment by olifolkerd** [comment]
That is happening because you havent set the ***height*** property on the table so you are not using the virtual DOM which means the table is trying to render all the rows at once.

To enable the Virtual DOM rendering mode you must set the ***height*** property to any valid css value.

For more information, checkout [The Documentation](http://tabulator.info/docs/3.3#virtual-dom)

Cheers

Oli

**Comment by tomheaps** [comment]
I only have 36 rows in the table which fit into the height of a single screen so virtual DOM shouldn't really be required should it?

**Comment by olifolkerd** [comment]
With 36 rows you should be fine, but i would generally recommend using it as it leads to a smoother experience all round.

There is no way that Tabulator should be consuming that much memory with 36 rows, It should be a few meg at most. How are you determining the memory usage?

Cheers

Oli

**Comment by tomheaps** [comment]
Solved the problems with memory bloat/leaking with my page.

Before you helped me to fix the updateData issue, I was using setData to update table data from a MySQL database. I was doing an ajax call to the database every 60 seconds to update.

Each time the call to setData was made, the memory consumed by my webpage increased by 5-6 MB, causing it to crash within an hour of it being open.

I think the problem is that the table cells being replaced by setData are not being destroyed by tabulator and instead they float around as detached DOM elements. They won't ever be removed by browser garbage collection because most of them are still referenced by event handlers as they have things like cellClick associated with them and thousands of javascript objects will accumulate in a short space of time.

**Comment by olifolkerd** [comment]
I will look at improving that  in the next release

**Comment by olifolkerd** [comment]
Hey All,

You will be happy to hear i have just pushed a fix for the memory leak issue to the 3.4 branch which will be released tomorrow.

Cheers

Oli :)

**Comment by olifolkerd** [comment]
Tabulator version 3.4 has now been officially released!

You can find details of the new release in the [News Section](http://tabulator.info/news)

Full documentation of all the new features can be found [Here](http://tabulator.info/docs/3.4)

Cheers

Oli :)
